### ISSUE TITLE
api.tools.parse_actions doesn't work with new gmail interface

### ISSUE THREAD

**Issue by OlegWock** [KartikTalwar/gmail.js#467]
I noticed that observer doesn't work with new GMail interface. I added some `console.log` and noticed that responses are in format, different from expected in `api.tools.parse_actions`. Could somebody fix that? I tried to decode Gmail responses myself, but with no result, they are very ugly and unreadable

**Comment by josteink** [comment]
New Gmail not being fully supported? [It is known.](https://media.giphy.com/media/10hexADl3Kl6SY/giphy.gif)

New Gmail breaks quite a lot of things. Fully supporting new Gmail is a ongoing effort.

Some code has already been committed and release in 0.7.0 to ensure a minimal amount of compatibility (like not *crashing*).

Right now event-observers are on top of my priority-list too, since that's the only remaining thing broken in my extension (after the other updates I've applied).

Hopefully more fixes and commits will come as we reverse engineer some of the new XHR formats used.

If you find some other things you depend on not working, I advice you to try to resolve the issue yourself, and submit PRs to get your patches mainlined.

**Comment by josteink** [comment]
Some new commits have been added and a new package released (0.7.1). 

Notable changes:

* preliminary code for parsing new XHR-format added. 
* less crashing/errors in XHR interceptor
* send_message event should now trigger.

**Comment by ediAm** [comment]
Hi,
Can you post a link for the 0.7.1 package release please ?
I can't find it in the code branches, i found only 0.6.4

Thank you very much !

**Comment by josteink** [comment]
> Hi,
> Can you post a link for the 0.7.1 package release please ?
> I can't find it in the code branches, i found only 0.6.4

We don't really tag releases in source-code. If you need a specific version, check for when we bump-versions (post release) in `package,json`.

Currently, as of right now, 0.7.1 is git master.

**Comment by phyxolog** [comment]
> send_message event should now trigger.

I have an error when use send_message observers.

![screenshotfrom2018-05-0719-05-24](https://user-images.githubusercontent.com/5149171/39711976-18f3abc8-522a-11e8-8287-9416c09ef22c.jpg)

**Comment by phyxolog** [comment]
@josteink Look, this error called every time, when i'm use Gmail.JS on new Gmail UI.
'Before' event on send_message work fine. But to 'after' event not reached (break on this error in debugger).

Error on this url: https://mail.google.com/sync/u/1/i/s?hl=ru&c=42
Maybe because request payload on this url look like a:
`2=[object Object]&3=[object Object]&4=[object Object]&5=2`

![screenshot from 2018-05-08 12-59-12](https://user-images.githubusercontent.com/5149171/39751041-a63959ac-52bf-11e8-9768-2f914b4489eb.png)

**Comment by josteink** [comment]
I'm using this event as `api.observe.after("send_message") in one of my production extensions and it works flawlessly. That's why I'm asking for a minimal repo to reproduce the error.

**Comment by josteink** [comment]
I'm using this event as `gmail.observe.after("send_message")` in one of my production extensions and it works flawlessly, also on new Gmail. That's why I'm asking for a minimal repo to reproduce the error.

That will help us determine what kind of error it is, and how to best fix it.

**Comment by phyxolog** [comment]
@josteink I think problem in this line:
`body = arguments[0] = this.xhrParams.body_is_object ? this.xhrParams.body_params : $.param(this.xhrParams.body_params,true).replace(/\+/g, "%20");`

Strangely, but when `typeof body_params == "object"` -  `body_is_object == undefined`.
I'm added this fix to library (before set `body` in before trigger):

`if (this.xhrParams.body_is_object === undefined && typeof this.xhrParams.body_params === "object") {
  this.xhrParams.body_is_object = true;
}`

and worked. But i'm not sure, it's right or wrong.

Minimal repo - it's include gmail.js to page and set observers. So, I don't see the point in this.

**Comment by phyxolog** [comment]
@josteink you used anywhere `gmail.observe.before`?
I'm was wrong. `Before` trigger don't work fine.
In `api.observe.trigger` at 1246 line we look this code:

```javascript
body = arguments[0] = this.xhrParams.body_is_object ? this.xhrParams.body_params : $.param(this.xhrParams.body_params,true).replace(/\+/g, "%20");
```

When Gmail make request to `https://mail.google.com/sync/u/1/i/s?hl=ru&c=x`, this line work wrong.
It's transform this object:
![screenshot from 2018-05-08 16-41-45](https://user-images.githubusercontent.com/5149171/39760710-be68c660-52de-11e8-814e-71ee86836505.png)
to this string: `2=[object Object]&3=[object Object]&4=[object Object]&5=2`.
But we see in network inspector, what payload should be:
![screenshot from 2018-05-08 16-42-40](https://user-images.githubusercontent.com/5149171/39760891-42e61f28-52df-11e8-9d52-6f536dc6810b.png)
It happens when i'm have this code in my extension:
```javascript
this.gmail.observe.before("send_message", (url, body, data, xhr) => {});
```

**Comment by josteink** [comment]
I haven’t used the ‘before’-event, so that might need some fixing. 

Could you doublecheck if you still have this error on version 0.7.2?

**Comment by phyxolog** [comment]
@josteink Yes, i have error on 0.7.2.

**Comment by josteink** [comment]
I’ll see if I can take a look at this when I’m back at work Monday.

**Comment by phyxolog** [comment]
@josteink i'm not sure, but i solved this problem with this code:

````javascript
if (api.check.is_new_gui()) {
  body = arguments[0] = this.xhrParams.body_is_object ? this.xhrParams.body_params : JSON.stringify(this.xhrParams.body_params);
} else {
  body = arguments[0] = this.xhrParams.body_is_object ? this.xhrParams.body_params : $.param(this.xhrParams.body_params,true).replace(/\+/g, "%20");
}
````
Because new request needed a payload, not a query string parameters.
Maybe it's works only with send_message, i'm not tested with other interceptors.

**Comment by josteink** [comment]
Thanks for the suggestion and fix!

I've tested it. It runs fine and looks good, but there's a few nuances you're not properly handling, which I think I would suggest you also take care.

* `api.check.is_new_gui()` is purely about the new *GUI*, no the new, underlying *data-layer*.
* Google has said they will be deploying the new data layer independent of the new GUI (that is, users may be getting the new data-layer without the new GUI).
* This is not related to UI, but related to the new data-contract.
* Therefore you probably want to check `api.check.is_new_data_layer()` instead.

Since this is your fix, you deserve the credit (and to be [listed among the contributors](https://github.com/KartikTalwar/gmail.js/graphs/contributors)), so I would really appreciate if you could send this as a PR I could merge!

Could you do that?

**Comment by phyxolog** [comment]
@josteink Yes, i'm sended a PR #474.

**Comment by josteink** [comment]
Closed via #474.
